sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala

```diff
@@ -255,6 +274,35 @@ class StateStoreChangelogWriterV2(
   }
 }
 
+class StateStoreChangelogWriterV3(
+     fm: CheckpointFileManager,
+     file: Path,
+     compressionCodec: CompressionCodec,
+     stateStoreCheckpointIdLineage: Array[LineageItem])
+  extends StateStoreChangelogWriterV1(fm, file, compressionCodec) {
+
+  override def version: Short = 3
+
+  writeVersion()
+  // Also write lineage information to the changelog, it should appear
+  // in the second line for v3 because the first line is the version
+  writeLineage(stateStoreCheckpointIdLineage)
+}
+
+class StateStoreChangelogWriterV4(
```

Comment:
scaladoc please. Why is this a new version? How is the encoding different

Reply:
There is a `merge ` operation in V2 reader / writer, also the eof token is different